### PR TITLE
NameObject wrong encoding

### PR DESCRIPTION
**Pull request by Rak424**
Name object  encodes all characters from 1 to 255, so it's not correct to use utf-8.

```
7.3.5 Name objects
Beginning with PDF 1.2 a name object is an atomic symbol uniquely defined by a sequence of any characters (8-bit values) except null (character code 0).
```

### CODE REVIEW

**Comment by pubpub-zz** [comment]
@Rak424 
on page 18 of PDF32000-1:2008 you will see
image

The 8bits only encoding will be done with Latin1

**Comment by pubpub-zz** [comment]
@Rak424 
please provide a pdf with the issue you are trying to fix. without I can not consider your PR as acceptable

**Comment by art049** [comment]
Hey @Rak424 and @pubpub-zz, We've been running the benches on this branch, and it seems these changes are significantly improving the `PDFWriter` performance. I think it's worth considering in merging this or not.

image

To obtain those, we installed CodSpeed on [a fork synced with this repo](https://github.com/CodSpeedHQ/pypdf/pull/17) You can have a look at the full report [here](https://codspeed.io/CodSpeedHQ/pypdf/branches/name).

File: tests/test_generic.py [hunk 1]
```diff
@@ -266,11 +249,29 @@ def test_name_object(caplog):
     with pytest.raises(PyPdfError):
         NameObject("/\0").write_to_stream(BytesIO())
 
+    # testing all allowed values
+    NameObject.encoding = "latin1"
```
Comment (stefan6419846) [inline tests/test_generic.py:252]: Have you verified that this has no side-effects if the test fails?

**Comment by pubpub-zz** [comment]
> Hey @Rak424 and @pubpub-zz, We've been running the benches on this branch, and it seems these changes are significantly improving the `PDFWriter` performance. I think it's worth considering in merging this or not.
> 
> image
> 
> To obtain those, we installed CodSpeed on [a fork synced with this repo](https://github.com/CodSpeedHQ/pypdf/pull/17). You can have a look at the full report [here](https://codspeed.io/CodSpeedHQ/pypdf/branches/name).
> 

The title identifies that this PR is dealing with an mi

**Comment by art049** [comment]
> If I read the detailed results correctly, this is a 4% difference on 1s. Can we really consider this a "significant improvement"?

It may seem small, but with over 4.5M downloads per month, I think the cumulated impact of such a change is significant.

> The title identifies that this PR is dealing with an invalid handling. I'm concerned about getting the good result and preventing regressions.
Are you dealing with decoding? Encoding?

Yes maybe this speedup means there is a regression somewhere. Overall the bench includes encoding afaik:
image

File: tests/test_generic.py (hunk 1 above)
Comment (Rak424) [reply tests/test_generic.py:252]: null char is explicitly forbidden in specs, I can not guarantee that's there is no side effects, but risk seems very low since it's the end of string symbol used in C, so using it could have much more side effects in other implementations. And the risk is only in write_to_stream, not in parsing, furthermore it will create a corrupted file to allow it.

Comment (stefan6419846) [reply tests/test_generic.py:252]: This is not about the null character, but IMHO you are changing the class default here without ensuring that a test failure will revert this value correctly without affecting all the other tests.

**Comment by pubpub-zz** [comment]
> Sorry to not explain directly what I've done, but I've seen different problems to solve in that part of code and some cases where clearly badly handled, here is a simple example, but not the only one:
> 
> ```
> bio = BytesIO()
> NameObject.read_from_stream(BytesIO(b"/#FF"), None).write_to_stream(bio)
> print(bio.getvalue)
> > b'/#C3#BF'
> ```

`b'/\xc3\xbf'.decode("utf-8") ` returns properly ÿ which is equivalent to `b'/\xff'.decode("latin-1") `

for me there is no error.


> 
> The way it was done was not optimal, so I've decided to try to refactor it and I'm not surprised about perf improvement.
> 
> A correct implementation should be able to do something like this, since from specs encoding is not defined on Name object itself, but I know that's not easy to make that change:
> 
> ```
> NameObject(b"hello")
> NameObject(b'\xe4\xbd\xa0\xe5\xa5\xbd\xe4\xb8\x96\xe7\x95\x8c (%)')
> ```
> 
> utf-8 is only for the specific case of creating a name from a text, but doesn't allow to handle all possible values, so I've added a fallback with latin1 who handle all values on a byte.
> 
> I often need some days/week and multiple tries to reach an optimal solution, tell me if you see some problems or changes to do.

remember thatthe current code provides with NameObject.CHARSETS a way to add / remove charsets from the possible solutions and the default, based on other PRs uses the utf-8, gbk and then latin-1. all these cases come from actual PDFs. have you experienced any actual issues with some PDF?

**Comment by Rak424** [comment]
`b"/#FF"` corresponds to value `b'\xff'` and  `b'/#C3#BF'` to value `b'\xc3\xbf'`, that s not the same value and there is no encoding notion here, specs only says that in the specific case of when you try to use string value, you have to encode it with utf8, but utf8 doesn't allow to use all chars from 1 to 255, and current implementation of name object only allows str and not bytes in constructor, so my workaround is to fallback to  an encoding who allows all values in one byte, latin1 seems a good choice for me. I don t know why gbk is present here but it seems useless. I don't have an example to show but the current implementation doesn't respect specifications, I'm not talking about handling wrong pdf. Don't hesitate to tell me if I've misunderstood something, pdf format is very very very bad.

**Comment by pubpub-zz** [comment]
> `b"/#FF"` corresponds to value `b'\xff'` and `b'/#C3#BF'` to value `b'\xc3\xbf'`, that s not the same value and there is no encoding notion here, specs only says that in the specific case of when you try to use string value, 

you have to encode it with utf8, but utf8 doesn't allow to use all chars from 1 to 255and current implementation of name object only allows str and not bytes in constructor.

NameObjects are actually string like : the proof is about using the encoding  with the '#' character. The byte encoding is just a matter of file encoding.

 so my workaround is to fallback to an encoding who allows all values in one byte, latin1 seems a good choice for me. I don t know why gbk is present here but it seems useless. 

This fix was introduced years ago. I' trying some archeology to find it

I don't have an example to show but the current implementation doesn't respect specifications, I'm not talking about handling wrong pdf. 

Don't hesitate to tell me if I've misunderstood something, pdf format is very very very bad.

I agree that the current implementation does not match the specification:
image

The historical approach in pypdf is more permissive and as we are rewriting all the names with PdfWriter, there is no issue with readers.
Moving to a full compliant solution may be introduce lots of regression that will be very tricky to detect (NameObjects are a sort of lazy linking)
As long as this PR does not fix an existing issue I'm not fond of it.

@MartinThoma  /  @MasterOdin  / @stefan6419846 
can you give your opinions?

**Comment by stefan6419846** [comment]
I have to admit that it is hard to follow this for me. If I understand it correctly, there are some different interpretations of the PDF specs regarding name objects.

In my understanding, names are just an arbitrary list of bytes. An application has to ensure a consistent handling of them for matching purposes, but as they are mostly irrelevant to the user and primarily about the internal structure, there are no further rules except for predefined sequences, for example as part of dictionary keys.

The newly introduced test sequences from ISO 32000-2:2020, Table 4 pass with the old implementation as well, thus I do not see any real need for changing anything about this unless there is a real example where the current implementation fails and the new one does not.

File: tests/test_generic.py [hunk 2]
```diff
@@ -180,64 +179,71 @@ def test_read_string_from_stream_excape_digit2():
...
+            b"6#E7#E8#E9#EA#EB#EC#ED#EE#EF#F0#F1#F2#F3#F4#F5#F6#F7#F8#F9#FA#FB#FC#FD#FE#FF",
+            bytes(range(1, 0x100)).decode("latin1"),
+        ),
+        (b"/\x80\x02\x03", "\x80\x02\x03", b"/#80#02#03"),
+    ):
+        name = NameObject.read_from_stream(BytesIO(test[0]), None)
+        assert name == f"/{test[1]}"
+        bio = BytesIO()
+        name.write_to_stream(bio)
+        assert bio.getvalue() == (test[0] if len(test) == 2 else test[2])
```
Comment (stefan6419846) [inline tests/test_generic.py:243]: Conditions inside tests are usually considered bad practice.

**Comment by Rak424** [comment]
to be more clear, here is two examples that can lead to problems hard to detect:

1. name value is modified after reading/writing operations:
```
bio = BytesIO()
NameObject.read_from_stream(BytesIO(b"/DocuSign#AE"), None).write_to_stream(bio)
print(bio.getvalue())  # > b'/DocuSign#C2#AE' != b"/DocuSign#AE"
```

2. all delimiter characters should be escaped, but it's not the case:
```
bio = BytesIO()
NameObject("/#()<>[]{}/%").write_to_stream(bio)
print(bio.getvalue())  # > b'/#23#28#29<>[]{}#2F#25'
```

**Comment by pubpub-zz** [comment]
@Rak424 can you provide actual PDF and code that shows some issues and not unitary.
About the two examples you are providing, 
1. even if the binary encoding are different, all names will be encoded in the same way and should match
2. the PDF specification states:
>"Regular characters that are outside the range EXCLAMATION MARK(21h) (!) to TILDE (7Eh) (~) **should** be written using the hexadecimal notation."
   the current implementation is valid

**Comment by Rak424** [comment]
hi

about example 2, from spec:
```
c) Any character that is not a regular character shall be written using its 2-digit hexadecimal code, preceded by the NUMBER SIGN only.
```
```
All characters except the white-space characters and delimiters are referred to as regular characters.
```
It's not a huge problem since it's easy to any parser to handle it, but could be a problem with an extremely strict parser.

The main problem is on example 1:
The original value is b'DocuSign\xae' who has no UTF8 representation and corresponds to 'DocuSign®' in latin1
The final value is  b'DocuSign\xc2x\ae' who corresponds to 'DocuSign®' in utf8 and 'DocuSignÂ®' in latin1

So, it means that if you append a pdf, it will creates a modification on these kinds of values. On a signed pdf where you add revocation infos, it will create a modification even if you don't modify the document.

**Comment by pubpub-zz** [comment]
@Rak424 
I see your example. However both  b'DocuSign\xc2\xae' and  b'DocuSign\xae' are translated into 'DocuSign®' NameObject.
Again I agree that the 2 forms are binary different so could be considered as different whereas pypdf will consider them as identical, but I do not see actual use cases where both versions will be used in the same PDF.
If a document uses version 2, all instances will be translated for version 1 and the document will be properly read.

If we do a change to keep the bytes internally, b'DocuSign\xae' will be stored, but without changing the arguments, NameObject('DocuSign®') will get internally  b'DocuSign\xc2\xae' and matching will not work anymore.
This will require lots of change and will break the API and induce lots of side effects in all programs. This is not acceptable

up to now you have not reported high level PDF that are failing. without I would consider your report as a acceptable deviation of pypdf to the PDF specification

**Comment by Rak424** [comment]
do you understand that b'DocuSign\xc2\xae' is not b'DocuSignx\ae' ?

**Comment by pubpub-zz** [comment]
I do if you consider a full compliance to the spec. however as said I do not consider that both can be met at the same time. Also we need a way to build a NameObject from an `str` that can match one or the other, or both.

Please provide a PDF file showing some issue

**Comment by ztravis** [comment]
At the very least can we add the full set of delimiter characters to `NameObject.renumber_table`? It's missing curly, angle and square brackets.

**Comment by stefan6419846** [comment]
Closing this due to lack of response. If you are still interested in working on this and have an actual use case where this causes issues, please report an issue first and propose a clean PR afterwards.
